### PR TITLE
Add TaskState to worker.py

### PR DESCRIPTION
**Pull request by gforsyth**
This is some initial progress towards #4097 -- dependents, dependencies and dep_state are still being tracked in a single dict on the Worker class but those come next.

Still lots to do on this, so no need for anyone to review just yet.

### CODE REVIEW

**Comment by mrocklin** [comment]
Hooray for PRs with incremental progress!

FWIW I really like the approach of starting with just the task-wise state.  That seems sanity-preserving.

**Comment by mrocklin** [comment]
Checking in here @gforsyth .  How are things going?  Have you run into tricky bits that I can help with?

**Comment by gforsyth** [comment]
> How are things going?

Definitely hit some tricky bits, but there's something I thought of I want to try.  
Mostly working, just some bookkeeping issues w.r.t. updating `data_needed`, `waiting_for_data`, etc. in a sane way.

I'll ping if the next few hours don't resolve things.

**Comment by mrocklin** [comment]
It's cool seeing all of the activity here :)  

How are things going for you @gforsyth ?

**Comment by gforsyth** [comment]
> How are things going for you @gforsyth ?

I think I'm closing in on a decent first cut of this.  I have a few more test failures I'm trying to trace out.

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py:1611]: I'm curious.  What caused the need for this?

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py]: I'm curious, what cause the need for this?

**Review by mrocklin — COMMENTED** [review]
Some minor comments. I realized that I had some other comments queued up here from a previous review that I neglected to send.  It looks like a few of those are obsolete now.

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:1611]: I think this is actually a result of some bad bookkeeping, but I stuck it in there to both give me a place to set a breakpoint, but also to see where else things broke if there wasn't a `KeyError` in `self._transitions`.  It should not be required once I've sorted that out.

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:1446]: ahh, yeah, this was a previous bit of disambiguation -- elsewhere I've stuck to being a little more verbose and saying `dependent` and `dependency`, I'll also make that change here.

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py]: This shouldn't happen anymore.  It was possible for a task to be instantiated with a default state of `None`, but I think setting it to `"waiting"` at instantiation nixes this issue.

**Comment by mrocklin** [comment]
Ooh, nice green check marks :heavy_check_mark:

**Comment by mrocklin** [comment]
@gforsyth are we ready to remove the WIP label from this PR?  Would it be reasonable to merge it?

**Comment by gforsyth** [comment]
> @gforsyth are we ready to remove the WIP label from this PR?  Would it be reasonable to merge it?

I can remove the WIP. I think it's ready for an initial review. There are comments to clean up and some docstrings to expand a bit. 

I think there's more that _could_ be done, but probably not much more that _needs_ to be done. And given that I'm going to try out speculative task assignment it might make more sense to merge this in closer to as-is (pending review) than to get it "perfect" before tearing it apart again.

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py:1550]: If the value is large enough to be placed into slow storage immediately (unlikely) then this might not be free.  We might want to drop it.

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py:1680]: Is it possible to reinstate this check?

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py:1679]: Tiny nit, but maybe `.clear()`?

If this was in the the scheduler I would recommend this due to what I would expect to be a small performance gain.  It probably doesn't matter as much here though.

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py:1779]: I would think that TaskState objects always have a `.state` attribute

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py:1803]: Should this be `dep in self.tasks` now?

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py:1828]: Is it safe to do `self.tasks[d]` here, or should we be concerned about `d not in self.tasks` as a possibility?

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py:1929]: Do we still need this `getattr` call or can we assume this?

Oh, or is this for if `d not in self.tasks`?

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py:1942]: I support docstrings generally, but when we add them we should probably adhere to numpdoc standard.  

I know that you know this stuff better than 99.999999% of the population (probably an underestimate) but I'll do what I always do, and copy in this doc: https://docs.dask.org/en/latest/develop.html#docstrings

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py:1957]: This really makes you want variable scope in list comprehensions :)

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py:2027]: ??  What happened here?

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py:2420]: Thanks for taking the time to encode this explanation

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py:2925]: Just checking, can you verify that these checks were moved to the other validate methods?

Most of our testing confidence comes from running these methods while operating Dask in different situations.  It's very important to us that we not lose validation here.

**Review by mrocklin — COMMENTED** [review]
A bunch of tiny comments!

**Comment by mrocklin** [comment]
Also, kudos on adding capability while dropping line-count :)

**Comment by mrocklin** [comment]
(unless that line count was due to dropping validation functions, in which case you will be hunted down by future generations of Dask maintainers)

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:1680]: Yep!

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:1679]: Sounds good.

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:1803]: Yes, I think so.  Thanks

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:1779]: yeah, I think this was from an earlier debugging pass, cleaning it up now.

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:1942]: Yep, these were notes to myself that I forgot to remove, but I'll make it a proper docstring instead (notes for my future self)

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:1957]: YES

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:1828]: I think it should be safe to do `self.tasks[d]` here, since the dependencies from `dep` should be created at this point.  I think I was worried about something dying in flight but that doesn't apply here.

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:1929]: > Oh, or is this for if `d not in self.tasks`?

Yeah, I should add a comment about that

Comment (gforsyth) [reply distributed/worker.py]: Or just make it more explicit in the first place

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:2925]: Thanks for the call out.  Deps are gone now, but all of these checks (except for `flight`) are covered in `validate_task_state`.  I'm adding an additional check for `flight` now, which should then cover all of these bases.

**Comment by gforsyth** [comment]
Thank you for the many small notes, @mrocklin.  It is my hope and belief that the line-count drop is mostly due to combining various validation functions, rather than removing them.  :crossed_fingers: that I am not the subject of the most dangerous game.

I think I've addressed all of your comments, but I'll take another cleaning pass now, too.

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py]: I genuinely don't remember why I thought this was necessary but tests pass locally when I remove it.

File: distributed/worker.py
Comment (mrocklin) [inline distributed/worker.py:2507]: Why does this need to be here?  What happens if it is removed?

Grepping through the codebase I see three paths for this:

1.  The task has been released during execution.  Maybe the user no longer needs the result
2.  The task is long-running, such as is called with `worker_client` or `secede`
3.  The same task is run twice (??)

In some of these situations I would expect the following `ts.state not in ("executing", "long-running")` check to clean things up.

File: distributed/worker.py
Comment (jrbourbeau) [inline distributed/worker.py:2256]: It looks like we've replaced `release_dep` calls with `release_keys`. What was the reasoning for this? It'll impact worker plugins

**Review by jrbourbeau — COMMENTED** [review]
Apologies for the delayed review @gforsyth. Thanks for all your careful work here!

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:728]: Restoring that now, I had removed `waiting` entirely from the `Worker` class but restored it so we don't have to do a loop over all pending tasks when we update this

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:121]: Yeah, I didn't want to use the name `type`, even as an attribute.  Happy to change this if it reads weird.

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:1526]: Nice catch!  Thanks!

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:1584]: Definitely.  I thought I'd changed all of these but missed this one.  Thanks!

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:1800]: Yep, tested and confirmed this.

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:2256]: There's no longer a distinction between a `dep` and a `Task` which is why I removed it.  But if it's going to impact the plugins I can restore it and have it wrap `release_keys` (or we can update the plugins?)

File: distributed/worker.py
Comment (gforsyth) [reply distributed/worker.py:2511]: Hmm, 
I had put this in because there were tests in `test_failed_worker.py` that would throw assertion errors (but not fail the test in question) but I've just removed it and can no longer get those errors to show up, so it may have been due to a problem elsewhere that is now fixed.

File: distributed/worker.py
Comment (mrocklin) [reply distributed/worker.py:2256]: > It'll impact worker plugins

Do we know of any worker plugins that trigger on `release_dep`?

Also, If we're jumping from version 2 to 2020 in the next release then we can also break some things.  Breaking `WorkerPlugin.release_dep` sounds ok to me.

File: distributed/worker.py
Comment (jrbourbeau) [reply distributed/worker.py:121]: Gotcha. FWIW it reads a little weird to me, so I have a slight preference for switching to `type` instead of `type_` (similar to what we have in the [`FutureState` class](https://github.com/dask/distributed/blob/946c6a46abb04615051537f1bf39b81163aa91f0/distributed/client.py#L432))

File: distributed/worker.py
Comment (jrbourbeau) [reply distributed/worker.py:2256]: > Do we know of any worker plugins that trigger on release_dep?

None come to mind, but there may be user-defined plugins that use `release_dep`

> Breaking WorkerPlugin.release_dep sounds ok to me

Yeah, I think we're treating all the HLG changes + the bump to calver as a major version change, so removing `release_dep` should be okay

**Review by jrbourbeau — COMMENTED** [review]
Thanks for the updates @gforsyth! It looks like there are a few remaining `executing` -> `executing_count` updates, but other this looks good to go

**Comment by mrocklin** [comment]
Passed tests here are encouraging.  I'm going to restart Github actions CI to see if this is consistent.

**Review by jrbourbeau — APPROVED** [review]
Thanks @gforsyth! I'm +1 on merging the changes here. I'll leave this open for a bit to allow for last minute feedback, but then I plan on merging

**Pull request merged**
